pilot/pkg/config/kube/gateway/deploymentcontroller.go

```diff
@@ -585,6 +582,20 @@ func (d *DeploymentController) apply(controller string, yml string) error {
 	// Ensure our canManage assertion is not stale
 	us.SetResourceVersion(resourceVersion)
 
+	// Because in 1.24 we removed old label "istio.io/gateway-name", in order to not mutate the deployment.spec.Selector during upgrade.
+	// we always use the old `selector` value
+	if gvr.Resource == "deployments" {
+		deployment := d.deployments.Get(us.GetName(), us.GetNamespace())
+		if deployment != nil && deployment.Spec.Selector.MatchLabels["istio.io/gateway-name"] != "" {
+			us.Object["spec"].(map[string]any)["selector"] = deployment.Spec.Selector
+			us.Object["spec"].(map[string]any)["template"].(map[string]any)["metadata"].(map[string]any)["labels"] = deployment.Spec.Template.Labels
```

Comment:
This is basically saying "Whatever the labels on the existing deployment are is now the labels"

This means if a label was added, for instance, it would be ignored.

I think we do not want this, we only want to _add_ "istio.io/gateway-name"

Reply:
I think  only adding "istio.io/gateway-name" is not enough, say one day we may add/remove another label.

Reply:
NVM, We cannot add labels by patching

Reply:
ping @howardjohn We cannot either add/remove a label

Reply:
Yes we can. The metadata labels is dynamic and controlled by what is on the Gateway. It's only the selector that is immutable. We need to add the legacy label to metadata and change the selector if they were previously using the old selector, but only *add* it not remove it.